src/watch.rs

```diff
@@ -71,33 +100,75 @@ struct FilteredOut<'a> {
     path: PathBuf,
 }
 
-impl Watch {
-    /// Instantiate a new Watch.
-    pub fn new(logger: slog::Logger) -> Result<Watch, notify::Error> {
-        let (tx, rx) = chan::unbounded();
+struct Filter {
+    notify: RecommendedWatcher,
+    notify_rx: Receiver<notify::Result<notify::Event>>,
+    watch_rx: Receiver<Vec<WatchPathBuf>>,
+    event_tx: Sender<Vec<PathBuf>>,
+    watches: HashSet<PathBuf>,
+    logger: slog::Logger,
+}
 
-        Ok(Watch {
-            notify: Watcher::new(tx, Duration::from_millis(100))?,
-            watches: HashSet::new(),
-            rx,
-            logger,
-        })
+impl Filter {
+    fn run(mut self) {
+        loop {
+            select! {
+                recv(self.notify_rx) -> msg => match msg {
+                    Ok(event) => if let Some(paths) = self.process_watch_events(event) {
+                        let paths = self.buffer_notify(paths);
+                        debug!(self.logger, "notification on watched paths"; "paths" => ?paths);
+                        if let Err(e) = self.event_tx.send(paths) {
+                            debug!(self.logger, "send error"; "error" => ?e)
+                        }
+                    },
+                    Err(chan::RecvError) => {
+                        debug!(self.logger, "filesystem notify channel was disconnected");
+                        return
+                    }
+                },
+                recv(self.watch_rx) -> msg => match msg {
+                    Ok(paths) => {
+                        let path_log = format!("{:?}", paths);
+                        if let Err(e) = self.extend(paths) {
+                                debug!(self.logger, "error extending watch paths:"; "error" => ?e, "paths" => path_log)
+                        }
+                    },
+                    Err(chan::RecvError) => {
+                        debug!(self.logger, "watch extension channel was disconnected");
+                        return
+                    }
+                }
+            }
+        }
+    }
+
+    fn buffer_notify(&self, mut touched: Vec<PathBuf>) -> Vec<PathBuf> {
+        let until = Instant::now() + Duration::from_millis(200);
+        loop {
+            match self.notify_rx.recv_deadline(until) {
+                Ok(event) => {
+                    if let Some(mut paths) = self.process_watch_events(event) {
+                        touched.append(&mut paths);
+                    }
+                }
+                Err(_) => {
+                    touched.dedup();
+                    return touched;
+                }
+            }
```

Comment:
I don’t think it’s a good idea to have a loop-in-a-loop that only watches for one message. The outer message `select!` will not be live while this blocks.

Reply:
We already listen for filesystem_events_rx outside of this helper, so blocking *only* on that means we don’t get any messages for e.g. new filepaths to watch, which might run over the channel bounds of those channels.

Reply:
We're watching on a deadline here, so this is where the 100ms goes to live in this code. Effectively, we're debouncing now at the watched path level, rather than on an Event level, which is what was required to solve the "vim-wait" test (although I think that covers a more general use case, of any complex file manipulation.)

Indeed, watching new paths would be the only messages that have to wait while we debounce, and even in the perverse case where many new paths were registered in 100ms, the channel is unbounded. 

There's a possible edgecase where a single path is reported as needing to be watched many times, so if we were consuming off the watch channel continuously, we could effectively deduplicate them as they were added to the HashMap, and in the meantime they consume memory in the channel. It seems like a particularly low risk issue though.

Reply:
I think the main issue I had with this is that, instead of consuming events as fast as they come in, we’d now add a fixed 200ms delay to our event loop. So if the user sends a new path to be watched, we’d not react for 200ms while in the inner loop.

Reply:
But check out the new code, it does not have that problem.